### PR TITLE
Create SSM Document Resource

### PR DESCRIPTION
**Pull request by gswai123**
This PR adds SSM Documents as a new resource to Custodian. It also includes the following:
- A cross-account filter for SSM docs which finds documents that are shared with non-whitelisted accounts
- A remove-sharing action which will remove account permissions for external accounts on an SSM document
- A delete action to delete SSM documents

Issue Link: https://github.com/cloud-custodian/cloud-custodian/issues/6575

### CODE REVIEW

File: c7n/resources/ssm.py [hunk 1]
```diff
@@ -597,3 +598,130 @@ def register_resource(cls, registry, resource_class):
...
+class SSMDocumentCrossAccount(CrossAccountAccessFilter):
+    """Filter SSM documents which have cross account permissions
+
+    :example:
+
+    .. code-block:: yaml
+
+            policies:
+              - name: ssm-cross-account
+                resource: ssm
```
Comment (JohnHillegass) [inline c7n/resources/ssm.py:626]:
```suggestion
                resource: ssm-document
```

File: c7n/resources/ssm.py [hunk 2]
```diff
@@ -597,3 +598,117 @@ def register_resource(cls, registry, resource_class):
...
+            delta_accounts = shared_accounts.difference(self.accounts)
+            if delta_accounts:
+                r['c7n:CrossAccountViolations'] = list(delta_accounts)
+                results.append(r)
+        return results
+
+
+@SSMDocument.action_registry.register('remove-sharing')
+class RemoveSharingSSMDocument(Action):
+    """Remove external accounts from SSM document permissions
```
Comment (thisisshi) [inline c7n/resources/ssm.py:672]: If this actions is dependent on a previous filter then that needs to be called out/validated for and documented in a docstring

File: c7n/resources/ssm.py [hunk 3]
```diff
@@ -597,3 +598,117 @@ def register_resource(cls, registry, resource_class):
...
+    schema = type_schema('remove-sharing')
+    permissions = ('ssm:ModifyDocumentPermission',)
+
+    def process(self, resources):
+        client = local_session(self.manager.session_factory).client('ssm')
+        for r in resources:
+            client.modify_document_permission(
+                Name=r['Name'],
+                PermissionType='Share',
+                AccountIdsToRemove=r['c7n:CrossAccountViolations']
```
Comment (thisisshi) [inline c7n/resources/ssm.py:684]: would also be useful to support the ability to pass in specific account ids to remove

File: c7n/resources/ssm.py [hunk 4]
```diff
@@ -597,3 +598,117 @@ def register_resource(cls, registry, resource_class):
...
+
+    schema = type_schema('delete')
+    permissions = ('ssm:DeleteDocument', 'ssm:ModifyDocumentPermission')
+
+    def process(self, resources):
+        client = local_session(self.manager.session_factory).client('ssm')
+        for r in resources:
+            try:
+                client.delete_document(Name=r['Name'], Force=True)
+            except client.exceptions.InvalidDocumentOperation:
```
Comment (thisisshi) [inline c7n/resources/ssm.py:701]: all this should be put behind a `force` attribute on the action itself, this does a whole lot more than just delete the document and can be ambiguous to users

File: c7n/resources/ssm.py [hunk 5]
```diff
@@ -597,3 +598,117 @@ def register_resource(cls, registry, resource_class):
...
+            )
+
+
+@SSMDocument.action_registry.register('delete')
+class DeleteSSMDocument(Action):
+    """Delete SSM documents
+    """
+
+    schema = type_schema('delete')
+    permissions = ('ssm:DeleteDocument', 'ssm:ModifyDocumentPermission')
```
Comment (thisisshi) [inline c7n/resources/ssm.py:694]: missing a comma at the end of `ssm:ModifyDocumentPermission`

File: c7n/resources/ssm.py [hunk 6]
```diff
@@ -597,3 +598,117 @@ def register_resource(cls, registry, resource_class):
...
 
 resources.subscribe(PostItem.register_resource)
+
+
+@resources.register('ssm-document')
+class SSMDocument(QueryResourceManager):
+
+    class resource_type(TypeInfo):
+        service = 'ssm'
+        enum_spec = ('list_documents', 'DocumentIdentifiers', None)
```
Comment (kapilt) [inline c7n/resources/ssm.py:608]: we should be defaulting to Owner=Self there's no governance constraint on first (amazon) or third party documents as their not manageable from the account the policy is executing in.

File: c7n/resources/ssm.py [hunk 7]
```diff
@@ -597,3 +598,117 @@ def register_resource(cls, registry, resource_class):
...
+        name = 'Name'
+        date = 'RegistrationDate'
+        arn_type = 'Document'
+
+    permissions = ('ssm:DescribeDocumentPermission', 'ssm:ListDocuments', )
```
[the remark this answers is not in the document]
Comment (gswai123) [reply c7n/resources/ssm.py:613]: Good catch, I'll remove that!

File: c7n/resources/ssm.py (hunk 2 above)
Comment (gswai123) [reply c7n/resources/ssm.py:672]: Sounds good, I just added to the docstring.

File: c7n/resources/ssm.py (hunk 4 above)
Comment (gswai123) [reply c7n/resources/ssm.py:701]: Hey Sonny! Could you explain what you mean by this a little bit more?

File: c7n/resources/ssm.py (hunk 5 above)
Comment (gswai123) [reply c7n/resources/ssm.py:694]: Fixed

File: c7n/resources/ssm.py (hunk 6 above)
Comment (gswai123) [reply c7n/resources/ssm.py:608]: Got it -- I just made this change.

File: c7n/resources/ssm.py (hunk 4 above)
Comment (thisisshi) [reply c7n/resources/ssm.py:701]: here you're doing additional actions (removing shared accounts) which is not the expected default behavior for a "delete" action, see example: iam-role.actions.delete

Comment (gswai123) [reply c7n/resources/ssm.py]: Got it -- I just pushed up some changes in which I added a "force" attribute to the delete action, and it won't modify document permissions if "force" isn't set to True.

File: c7n/resources/ssm.py [hunk 8]
```diff
@@ -597,3 +598,155 @@ def register_resource(cls, registry, resource_class):
...
+            shared_accounts = {
+                g.get('AccountId') for g in attrs}
+            delta_accounts = shared_accounts.difference(self.accounts)
+            if delta_accounts:
+                r['c7n:CrossAccountViolations'] = list(delta_accounts)
+                results.append(r)
+        return results
+
+
+@SSMDocument.action_registry.register('remove-sharing')
```
Comment (kapilt) [inline c7n/resources/ssm.py:671]: generally we want to do set-sharing to enable add/remove vs separate add/removes.. with a remove option supporting a special value of matched

Comment (gswai123) [reply c7n/resources/ssm.py:671]: Ah that makes sense. I'm working on this change now.

**Review by kapilt — APPROVED** [review]
thanks, lgtm

**Pull request merged**